subxt/src/client/offline_client.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::{
+    constants::ConstantsClient,
+    events::EventsClient,
+    rpc::RuntimeVersion,
+    storage::StorageClient,
+    tx::TxClient,
+    Config,
+    Metadata,
+};
+use derivative::Derivative;
+use std::sync::Arc;
+
+/// A trait representing a client that can perform
+/// offline-only actions.
+pub trait OfflineClientT<T: Config>: Clone + Send + Sync + 'static {
+    /// Return the provided [`Metadata`].
+    fn metadata(&self) -> Metadata;
+    /// Return the provided genesis hash.
+    fn genesis_hash(&self) -> T::Hash;
+    /// Return the provided [`RuntimeVersion`].
+    fn runtime_version(&self) -> RuntimeVersion;
+
+    /// Work with transactions.
+    fn tx(&self) -> TxClient<T, Self> {
+        TxClient::new(self.clone())
+    }
+
+    /// Work with events.
+    fn events(&self) -> EventsClient<T, Self> {
+        EventsClient::new(self.clone())
+    }
+
+    /// Work with storage.
+    fn storage(&self) -> StorageClient<T, Self> {
+        StorageClient::new(self.clone())
+    }
+
+    /// Access constants.
+    fn constants(&self) -> ConstantsClient<T, Self> {
+        ConstantsClient::new(self.clone())
+    }
+}
+
+/// A client that is capable of performing offline-only operations.
+/// Can be constructed as long as you can populate the required fields.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""), Clone(bound = ""))]
+pub struct OfflineClient<T: Config> {
+    inner: Arc<Inner<T>>,
+}
+
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""), Clone(bound = ""))]
+struct Inner<T: Config> {
+    genesis_hash: T::Hash,
+    runtime_version: RuntimeVersion,
+    metadata: Metadata,
+}
+
+impl<T: Config> OfflineClient<T> {
+    /// Construct a new [`OfflineClient`], providing
+    /// the necessary runtime and compile-time arguments.
+    pub fn new(
+        genesis_hash: T::Hash,
+        runtime_version: RuntimeVersion,
+        metadata: Metadata,
+    ) -> OfflineClient<T> {
+        OfflineClient {
+            inner: Arc::new(Inner {
+                genesis_hash,
+                runtime_version,
+                metadata,
+            }),
+        }
+    }
+
+    /// Return the genesis hash.
+    pub fn genesis_hash(&self) -> T::Hash {
+        self.inner.genesis_hash
+    }
+
+    /// Return the runtime version.
+    pub fn runtime_version(&self) -> RuntimeVersion {
+        self.inner.runtime_version.clone()
+    }
+
+    /// Return the [`Metadata`] used in this client.
+    pub fn metadata(&self) -> Metadata {
+        self.inner.metadata.clone()
```

Comment:
Might be better to return borrowed `&Metadata` since it can be quite big? More of an issue in resource constrained environments where `OfflineClient` is more likely to be used.

User can always manually clone if required.

Reply:
It's all behind an Arc anyway; the reason we return a clone is so that the OnlineClient (which has it behind an RwLock) can also meet this type signature (It couldn't return a reference from behidn the lock; it'd need to return a guard)